### PR TITLE
internal/charmstore: fix race in blobstore migration

### PR DESCRIPTION
**Pull request by rogpeppe**
We were reading from the entity and writing to it concurrently.
The race detector did in fact pick up the mistake, but we
hadn't run with the race detector enabled...

Also, adjust the logs a little so it's clear which entities are
being migrated, and log any errors when an invalid
hash is used with the GC.

### CODE REVIEW

**Review by fabricematrat — APPROVED** [review]
LGTM

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/charmstore/2155/

File: internal/blobstore/blobstore.go [hunk 1]
```diff
@@ -292,8 +292,9 @@ func NewRefs(n int) *Refs {
 // Add records that the given hash is referenced.
 // It ignores the hash if it's invalid.
 func (r *Refs) Add(hash string) {
-	data, ok := decodeHash(hash)
-	if !ok {
+	data, err := decodeHash(hash)
+	if err != nil {
+		logger.Errorf("cannot add bad hash %q: %v", hash, err)
```
Comment (mhilton) [inline internal/blobstore/blobstore.go:297]: This really shouldn't be an error I don't think. We use this with refs that are empty a lot prev5blobhashes that aren't specified for instance.

File: internal/charmstore/store.go [hunk 2]
```diff
@@ -431,6 +431,9 @@ func (s *Store) ensureIndexes() error {
 // deleting all blobs that have not been referenced since
 // the given time.
 func (s *Store) BlobStoreGC(before time.Time) error {
+	// BEWARE: if this code does not add all the relevant blob
```
Comment (mhilton) [inline internal/charmstore/store.go:434]: who is this comment for?

**Review by mhilton — APPROVED** [review]
LGTM with a few thoughts.

File: internal/blobstore/blobstore.go (hunk 1 above)
Comment (fabricematrat) [reply internal/blobstore/blobstore.go:297]: @mhilton We added a check to not use add when it's empty in here https://github.com/juju/charmstore/pull/764/files#diff-20be4fa0bfcb7416de93e699d65b2d2aR459

File: internal/charmstore/store.go (hunk 2 above)
Comment (rogpeppe) [reply internal/charmstore/store.go:434]: you!

Comment (rogpeppe) [reply internal/charmstore/store.go:434]: (just wanted something to say that this is important code)

**Comment by rogpeppe** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci-gce.jujugui.org:8080/job/charmstore-merge

**Pull request merged**